subxt/src/backend/utils.rs

```diff
@@ -0,0 +1,244 @@
+//! RPC utils.
+
+use super::{StreamOf, StreamOfResults};
+use crate::error::Error;
+use futures::future::BoxFuture;
+use futures::{ready, FutureExt, Stream, StreamExt};
+use std::{future::Future, pin::Pin, task::Poll};
+
+/// Resubscribe callback.
+type ResubscribeGetter<T> = Box<dyn FnMut() -> ResubscribeFuture<T> + Send>;
+
+/// Future that resolves to a subscription stream.
+type ResubscribeFuture<T> = Pin<Box<dyn Future<Output = Result<StreamOfResults<T>, Error>> + Send>>;
+
+pub(crate) enum PendingOrStream<T> {
+    Pending(BoxFuture<'static, Result<StreamOfResults<T>, Error>>),
+    Stream(StreamOfResults<T>),
+}
+
+/// Retry subscription.
+struct RetrySubscription<T> {
+    resubscribe: ResubscribeGetter<T>,
+    state: PendingOrStream<T>,
+}
+
+impl<T> std::marker::Unpin for RetrySubscription<T> {}
+
+impl<T> Stream for RetrySubscription<T> {
+    type Item = Result<T, Error>;
+
+    fn poll_next(
```

Comment:
I wonder whether we could code golf this to somehting more like:

```rust
fn poll_next(
    mut self: Pin<&mut Self>,
    cx: &mut std::task::Context<'_>,
) -> Poll<Option<Self::Item>> {
    loop {
        match self.state {
            PendingOrStream::Pending(ref mut fut) => match ready!(fut.poll_unpin(cx)) {
                Err(e) => {
                    if e.is_disconnected_will_reconnect() {
                        self.state = PendingOrStream::Pending((self.resubscribe)());
                    } else {
                        return Poll::Ready(Some(Err(e)));
                    }
                }
                Ok(stream) => {
                    self.state = PendingOrStream::Stream(stream);
                },
            },
            PendingOrStream::Stream(ref mut s) => match s.poll_next_unpin(cx) {
                Poll::Ready(Some(Err(e))) if e.is_disconnected_will_reconnect() => {
                    self.state = PendingOrStream::Pending((self.resubscribe)());
                }
                other => return other,
            },
        };
    }
}
```

I also made the future failing return the err instead of `Ok(None)`. Maybe we should have a `done` flag that we set on such an error so that it returns `None` afterwards instead of polling again after errors? Or maybe just let the underlying thing react however it wants and we `fuse()` this is we prefer.

Reply:
Good idea, I added this but my code is slightly more verbose because I can't mutate it when have reference to in the match

Reply:
Could it be possible to add `PendingOrStream::EmptyState` variant and do a `std::mem::replace` to circumvent that?

Reply:
I'll resolve this and have look at the code in utils.

I added another enum variant PendingOrStream::Closed